README.md
# MooshakExtensions
Extensions to Mooshak programming contest managment system


report/changes.txt
Para habilitar a devolução do relatório é necessário habilitar a execução do script <mooshak>/contrib/myreport.sh pelo mooshak.
Para isso é necessário chamar esse script a partir da função de análise do mooshak, que se encontra no ficheiro: 

<mooshak>/packages/classes/Submission.tcl

No final desta função, na linha 415, adicionar o código seguinte:

## Analyze a submission, records Result and Observations and produces a report
Operation Submission::analyze { {service 0} {language ""} } {
	variable Contest
#(...)

#(end row 415)ew code:
#N
	#Add report to submitted files that can be downloaded by user
	#If it is NOT a zip, if not DOES not return report
	exec contrib/myreport.sh ${_Self_}  ${_Self_}/$Program  ${_Self_}/$Report  $Contest/problems/$Problem
}


RARS/rarsext.sh
#!/bin/bash
# 
# This script unzips and compiles RISC-V assembly source code and run it.
#
#    a) Only one file must have *.asm extension
#
#    b) Other supporting files must have a different extension
#       it is suggested *.inc
#
#	 c) If Environment file exists for this problem:
#			- deletes all *.asm files in the unzipped source folder
#			- cp Environment to <unzipped>/client.asm
#
#    Call:
#		myrars.sh <mooshak dir> <submission_file> <extension> <environment>
#
#    mooshak compile:
#		/bin/bash myrars.sh   $home   $file  $extension   $environment 2>compile.err
#
#    mooshak execute:
#		/bin/bash run.sh 2> run.err 2>run.err
#
# 		Note1:  Needed to separate in compile and execute, because in execute there
#				is no access permitions to Environment file in problem folder
#
#		Note2:  If willing to detect errors when running scripts, redirect stderr
#				to files in the submission directory
#
#    cmd line:
#		myrars.sh  /home/mooshak/  src.zip  .zip  data/contests/AC2122/problems/A/Environment
#
#
# Helder Daniel						
# hdaniel@ualg.pt


#MacOsX
xtraZipDir=__MACOSX

#Get script parameters
home="$1"
file="$2"
extension="$3"
environment="$home/$4"

runnerScript="run.sh"
compiler="$home/contrib/rars_533d3c0.jar"

#unpacking/work directory
workdir="${file%.*}"
#echo wkdir: $workdir 

#client file to add if NOT null
clientFile="$workdir/client.asm"

#Unpack files only if needed
#if revaluating DO NOT unzip SOURCE files again
if [ ! -d "$workdir" ]; then
	#/bin/tar xmzf $file > /dev/null
	/usr/bin/unzip -o $file > /dev/null
fi

# IF $environment exists for this problem (<problem>/Environment)
#   delete all *.asm and 
#   cp $environment $clientFile
#
#Note could NOT do with cp, due to permissions
#but on POO2021 it works
if [ -f "$environment" ]; then
    rm $workdir/*.asm
   	cp "$environment" "$clientFile"
fi

#Find main
main=$(find $workdir -name *.asm)
#echo main: $main

#compile and run
#echo "/usr/bin/java -jar $compiler me $main 2> /dev/null" > $runnerScript
echo "/usr/bin/java -jar $compiler me $main" > $runnerScript
chmod 750 "$runnerScript"



java/javaext.sh
#!/bin/bash
# 
# This script compiles java source code and run it.
#
#    a) if sole java file compile it and run it (file must have class with main function)
#
#    b) if zipped java files uncompresses it and compile all java files
#         run the class where main is specified (only one class can have a main file)
#
#    c) if specified in context Environment a client program, copies it to submission directory 
#      i) if a main function exists in this file it is ran
#     ii) also aditional classes to form a small API can be implemented in this file
#
#    Call:
#
#          myjava.sh   <mooshak dir> <submission_file> <extension> <problem_environment>
#
#    mooshak:
#          
#/bin/bash myjava.sh   $home         $file             $extension  $environment
#
#
#    cmd line
#
#	   myjava.sh   /home/mooshak/   src.zip           .zip     data/contests/POO1920/problems/P3/Environment
#
#
#	To run junit5 Mooshak "contest language" must be configure to:
#				MaxCompFork	10 (or 100)
#				MaxExecFork	10 (or 100)
#
#	And uncomment junit5 compile an run lines below
#
# Helder Daniel						
# hdaniel@ualg.pt


xtraZipDir=__MACOSX
extraCodeName="MooshakTester"
javaExt="java"
runnerScript="run.sh"
dfltWorkdir="src"

#Change for desired mode
username="mooshak"   # instdir: as in /home/$username/contrib (...)
junit="5"	      # junit 4 or 5
jdk="/usr"          #"/usr" or for local "/home/$username/jdk-15.0.2"
#jdk="/home/$username/jdk-15.0.2"
 
#get cur directory name
function curDir {
	result=$(basename $(pwd))
	echo $result
}


#get last created directory name
function newestDir {
	#remove extra Zip dir (MACOSX)
	if [ -d "$xtraZipDir" ]; then
		rm -r $xtraZipDir
	fi
	#get newest dir entry data
	result=$(ls -lF --sort=time | grep / | head -n1)

	#extract dir name
	echo ${result##*' '} | cut -d/ -f1

}


#Find file with main function
function findMainClass {
local workdir=$1

	#get 1st macth only
	result=$(grep -r "void[ ]*main" $workdir | head -n1)
	echo $result | cut -d: -f1 | cut -d. -f1
}


#get submitted code and unpack it if needed
function getSubmission {
local extension=$1
local file=$2

case "$extension" in
	.java)
	#Avoid error in creating dir in reevaluation
	if [ ! -d $dfltWorkdir ]; then
		/bin/mkdir $dfltWorkdir
	fi
	/bin/cp $file $dfltWorkdir
	;;

	.gz)
	#/bin/echo "tar.gz"
	#Send output of unzip to /dev/null to avoid returning it to mooshak.
	#Mooshak requires no output or consider an error
	/bin/tar xmzf $file > /dev/null
	;;

	.zip)
	#/bin/echo "zip"
	#Send output of unzip to /dev/null to avoid returning it to mooshak.
	#Mooshak requires no output or consider an error
	/usr/bin/unzip -o $file > /dev/null
	;;

	*)
	exit 1;
	;;
esac
}


#Add extra source code if it is defined in problem environment (from Mooshak GUI)
function addExtraCode {
local environment=$1
local workdir=$2
local extraCodeName=$3
local javaExt=$4

if [ -f "$environment" ]; then
	extraCode="$workdir/$extraCodeName.$javaExt"

	#Store extra source code in working dir with file name <submission dir>.java
	cp "$environment" "$extraCode"

	#If root package of submission not the default: <src>, set package 
	#in extracode file to workdir (root package name of submission)
	if [ "$workdir" != "src" ]; then
		echo "package $workdir;" | cat - $extraCode > temp && mv temp $extraCode
	fi

	#Search for junit4 or 5 in "$extraCode" 
	#junit4: "org.junit.Test"
	#junit5: "org.junit.jupiter.api.Test"
	#This ways detect junit4 or 5
	junit=$(grep "org.junit" "$extraCode" | wc -l)
	if [ $junit -gt 0 ]; then
		echo "junit"
	else
		#Search for a class $extraCodeName which should have main()
		result=$(grep -w "class[ ]*$extraCodeName" $extraCode)
		echo $result
	fi
fi
}



######################
#
# Main prg
#
######################

#Get script parameters
home="$1"
file="$2"
extension="$3"
environment="$home/$4"

#if revaluating DO NOT recreate files
#search for a report
#if [ -f "1.html" ]; then
#        exit 0
#fi

#get submission directory
curdir=$(curDir)
fullcurdir=$(pwd)

#if revaluating DO NOT recreate only SOURCE files
#but recreate run.sh.
#This allows to test changing the $ENVIRONMET Extra source code
if [ ! -f "1.html" ]; then
	#Unpack files if needed
	getSubmission $extension $file
fi

#Find unpacking/work directory
workdir=$(newestDir)

#Find file with main in submitted code
mainClass=$(findMainClass $workdir)
mainClassSoloName=$(echo ${mainClass##*'/'})
mainClassName=$(echo $mainClass | cut -d/ -f2-)

#Add extra source code if it is defined in problem environment (from Mooshak GUI)
#If it has a main() function set to run it instead of submitted main() function
result=$(addExtraCode $environment $workdir $extraCodeName $javaExt)
if [ "$result" != "" ]; then

	if [ "$result" != "junit" ]; then
		mainClassName=$extraCodeName
	fi
fi


#compile
find $workdir -name *.java > sources.txt

#This way allows that anyone submit junit 4/5 files too
#junit 4
if [ "$junit" = "4" ]; then
	$jdk/bin/javac -J-Dfile.encoding=UTF-8 -cp .:/home/$username/contrib/junit-4.12.jar @sources.txt -Xlint:-unchecked
fi
#junit 5
#(NEED fork>0 to run in safeexec)
if [ "$junit" = "5" ]; then
	$jdk/bin/javac -J-Dfile.encoding=UTF-8 -cp .:/home/$username/contrib/junit-platform-console-standalone-1.6.0.jar @sources.txt -Xlint #:-unchecked
fi

rm sources.txt


#Create runner script $runnerScript

#For alone files in zip
if [ "$workdir" = "" ]; then
	workdir="."
fi

echo "#!/bin/bash" > "$runnerScript"
if [ "$result" != "junit" ]; then
	#echo "$jdk/bin/java -cp $workdir $package$mainClassName 2>&1" >> "$runnerScript"
	echo "$jdk/bin/java -cp $workdir $mainClassName" >> "$runnerScript"
else
	#junit 4
	#Remove test time (which is variable) and last line
	#Not needed in below soultion, cause just prints error lines
	if [ "$junit" = "4" ]; then
		echo "$jdk/bin/java -cp .:/home/$username/contrib/junit-4.12.jar:/home/$username/contrib/hamcrest-core-1.3.jar:$workdir org.junit.runner.JUnitCore MooshakTester | grep 'Mooshak\|expected' | grep '^[^0-9]' | cut -d' ' -f2- 2>&1" >> "$runnerScript"
	fi

	#junit 5
	#(NEED fork>0 to run in safeexec)
	if [ "$junit" = "5" ]; then
		echo "$jdk/bin/java -jar /home/$username/contrib/junit-platform-console-standalone-1.6.0.jar --class-path $fullcurdir/$workdir --scan-class-path 2>&1" >> "$runnerScript"
	fi
fi

chmod 750 "$runnerScript"




RARS/myrars20.sh
#!/bin/bash
# 
# This script unzips and compiles RISC-V assembly source code and run it.
#
#    a) Only one file must have *.asm extension
#
#    b) Other supporting files must have a different extension
#       it is suggested *.inc
#
#	 c) If Environment file exists for this problem:
#			- deletes all *.asm files in the unzipped source folder
#			- cp Environment to <unzipped>/client.asm
#
#    Call:
#		myrars.sh <mooshak dir> <submission_file> <extension> <environment>
#
#    mooshak compile:
#		/bin/bash myrars.sh   $home   $file  $extension   $environment 2>compile.err
#
#    mooshak execute:
#		/bin/bash run.sh 2> run.err 2>run.err
#
# 		Note1:  Needed to separate in compile and execute, because in execute there
#				is no access permitions to Environment file in problem folder
#
#		Note2:  If willing to detect errors when running scripts, redirect stderr
#				to files in the submission directory
#
#    cmd line:
#		myrars.sh  /home/mooshak/  src.zip  .zip  data/contests/AC2122/problems/A/Environment
#
#
# Helder Daniel						
# hdaniel@ualg.pt
# January 2022


#MacOsX
xtraZipDir=__MACOSX

#Get script parameters
home="$1"
file="$2"
extension="$3"
environment="$home/$4"

runnerScript="run.sh"
compiler="$home/contrib/rars_533d3c0.jar"

#unpacking/work directory
workdir="${file%.*}"
#echo wkdir: $workdir 

#client file to add if NOT null
clientFile="$workdir/client.asm"

#Unpack files only if needed
#if revaluating DO NOT unzip SOURCE files again
if [ ! -d "$workdir" ]; then
	#/bin/tar xmzf $file > /dev/null
	/usr/bin/unzip -o $file > /dev/null
fi

# IF $environment exists for this problem (<problem>/Environment)
#   delete all *.asm and 
#   cp $environment $clientFile
#
#Note could NOT do with cp, due to permissions
#but on POO2021 it works
if [ -f "$environment" ]; then
    rm $workdir/*.asm
   	cp "$environment" "$clientFile"
fi

#Find main
main=$(find $workdir -name *.asm)
#echo main: $main

#compile and run
#echo "/usr/bin/java -jar $compiler me $main 2> /dev/null" > $runnerScript
echo "/usr/bin/java -jar $compiler me $main" > $runnerScript
chmod 750 "$runnerScript"



report/myreport.sh
#!/bin/bash
# 
# This script prepares a report to be sent to the user 
# along with the source code (IF it is a zip file)
#
# Note:Excludes test folders started by H* from report
#
# Helder Daniel						
# hdaniel@ualg.pt
# March 2020

Submission="$1"
Program="$2"
Report="$3"
Problem="$4"

filter="[^H]*"   #Exclude test folder started by H* from report

hide=0	#Number of hidden test inputs
hidef=$(($hide+1))

ftype=$(file $Program | cut -d' ' -f2)

#If it is NOT a zip, DOES not return report
if [ "$ftype" == "Zip" ]; then
  chmod 600 $Program

  #Add report to zip
  /usr/bin/zip -uj  $Program  $Report

  #Add output error 
  /usr/bin/zip -uj  $Program  $Submission/*.err

  #Add tests inputs to zip
  cp -r $Problem/tests $Submission
  testFolders=$(find $Submission/tests/$filter -name input | sort)
  testFolders=$(echo $testFolders | rev | cut -d' ' -f$hidef- | rev)

  for i in $testFolders; do
    new=${Submission}/tests/$(echo $i | rev | cut -d / -f 2 | rev)input
    mv $i $new
  done

  zip -uj  $Program  $Submission/tests/*input
  rm -rf $Submission/tests

  chmod 400 $Program
else
	echo "Cannot add report to submission, submitted code is NOT a ZIP archive"
fi


report/reportext.sh
#!/bin/bash
# 
# This script prepares a report to be sent to the user 
# along with the source code (IF it is a zip file)
#
# Note:Excludes test folders started by H* from report
#
# Helder Daniel						
# hdaniel@ualg.pt


Submission="$1"
Program="$2"
Report="$3"
Problem="$4"

filter="[^H]*"   #Exclude test folder started by H* from report

hide=0	#Number of hidden test inputs
hidef=$(($hide+1))

ftype=$(file $Program | cut -d' ' -f2)

#If it is NOT a zip, DOES not return report
if [ "$ftype" == "Zip" ]; then
  chmod 600 $Program

  #Add report to zip
  /usr/bin/zip -uj  $Program  $Report

  #Add output error 
  /usr/bin/zip -uj  $Program  $Submission/*.err

  #Add tests inputs to zip
  cp -r $Problem/tests $Submission
  testFolders=$(find $Submission/tests/$filter -name input | sort)
  testFolders=$(echo $testFolders | rev | cut -d' ' -f$hidef- | rev)

  for i in $testFolders; do
    new=${Submission}/tests/$(echo $i | rev | cut -d / -f 2 | rev)input
    mv $i $new
  done

  zip -uj  $Program  $Submission/tests/*input
  rm -rf $Submission/tests

  chmod 400 $Program
else
	echo "Cannot add report to submission, submitted code is NOT a ZIP archive"
fi


ClientServer/csext.sh
#!/bin/bash
# 
# This script compiles C source code and run it.
#
#    a) if sole C file compile it and run it (file must have main function)
#
#    b) if zipped C files uncompresses it and compile all C files
#       run the generated a.out executable (only one file can have a main file)
#  #All source files must be in same dir
#  #The ma/usrin file must be the only function in its file.
#  #Nothing usefull for the program besides main should be in this file: not any other function, or data declarations!
#  #This is required for C
#
#    c) if specified in context Environment a client program, copies it to submission directory 
#      i) if a main function exists in this file it is ran (the submitted file with main is deleted)
#     ii) also aditional classes to form a small API can be implemented in this file
#
#   call:
#			myc.sh <compiler> <mooshak dir> <submission_file> <extension> <problem_environment>  <extra_flags>  
#
#	example:
#			myc.sh "/usr/bin/gcc -Wall -lm" /home/mooshak/ src.zip .zip data/contests/POO1920/problems/P3/Environment "-lpthread"
#
# Helder Daniel						
# hdaniel@ualg.pt


xtraZipDir=__MACOSX
extraCodeName="MooshakTester"
#to compile *.c and *.cpp
srcExt="c*"
exec="a.out"
runnerScript="run.sh"
dfltWorkdir="src"

#get cur directory name
function curDir {
	result=$(basename $(pwd))
	echo $result
}


#get last created directory name
function newestDir {
	#remove extra Zip dir (MACOSX)
	if [ -d "$xtraZipDir" ]; then
		rm -r $xtraZipDir
	fi
	#get newest dir entry data
	result=$(ls -lF --sort=time | grep / | head -n1)

	#extract dir name
	echo ${result##*' '} | cut -d/ -f1

}


#Find file with main function
function findMainFunctionFile {
local workdir=$1

	#get 1st match only
	result=$(grep -r "main[ ]*(" $workdir | head -n1)
	echo $result | cut -d: -f1
}


#get submitted code and unpack it if needed
function getSubmission {
local extension=$1
local file=$2

case "$extension" in
	.gz)
	#/bin/echo "tar.gz"
	#Send output of unzip to /dev/null to avoid returning it to mooshak.
	#Mooshak requires no output or consider an error
	/bin/tar xmzf $file > /dev/null
	;;

	.zip)
	#/bin/echo "zip"
	#Send output of unzip to /dev/null to avoid returning it to mooshak.
	#Mooshak requires no output or consider an error
	/usr/bin/unzip $file > /dev/null
	;;

	*)
	#/bin/echo "C/C++ .c or .cpp"
	#Avoid error in creating dir in reevaluation
	if [ ! -d $dfltWorkdir ]; then
		/bin/mkdir $dfltWorkdir
	fi
	/bin/cp $file $dfltWorkdir
	;;
esac

#Give write permissions so that can be deleted 
#on reevaluation
#chmod -R u+w *
}


#Add extra source code if it is defined in problem environment (from Mooshak GUI)
function addExtraCode {
local environment=$1
local workdir=$2
local extraCodeName=$3
local srcExt=$4

#Note: file $environment must not exist or have 0 size
#If exists and have blank chars must be cleanned with mooshak GUI or from cmd line
if [ -f "$environment" ]; then    #exist
	if [ -s "$environment" ]; then #not empty
		#Search for a file with mainextract extension (c or cpp) and remove it
		result=$(findMainFunctionFile $workdir)
		mainExt=$(echo $result | rev | cut -d. -f1 | rev)
		rm "$result"

		#Store extra source code in working dir with file name <submission dir>.$srcExt
		extraCode="$workdir/$extraCodeName.$mainExt"
		cp "$environment" "$extraCode"
	fi
fi
}



######################
#
# Main prg
#
######################

#Get script parameters
compiler="$1"
home="$2"
file="$3"
extension="$4"
environment="$home/$5"
extraflags="$6"

#get submission directory
curdir=$(curDir)

#if revaluating DO NOT recreate files
#search for a report
#if [ -f "1.html" ]; then
#        exit 0
#fi

#if revaluating DO NOT recreate only SOURCE files
#but recreate run.sh.
#This allows to test changing the $ENVIRONMET Extra source code
if [ ! -f "1.html" ]; then
	#Unpack files if needed
	getSubmission $extension $file
fi

#Find unpacking/work directory
workdir=$(newestDir)

#Add extra source code if it is defined in problem environment (from Mooshak GUI)
#If it has a main() function set to run it instead of submitted main() function
addExtraCode $environment $workdir $extraCodeName $srcExt

#compile
serverExec="server"
clientExec="client"
serverSrc=$serverExec.$srcExt
clientSrc=$clientExec.$srcExt
compileMode=0  # 2 client/server; else single *.c

#Search for server and client source code
if [ -f $workdir/$serverSrc -a -f $workdir/$clientSrc ]; then
	compileMode=2
	$compiler $workdir/$serverSrc -o $workdir/$serverExec $extraflags
	$compiler $workdir/$clientSrc -o $workdir/$clientExec $extraflags
else
	$compiler $workdir/*.$srcExt -o $workdir/$exec $extraflags
fi

#Create runner script $runnerScript
#determine if in default package or not
echo "#!/bin/bash" > "$runnerScript"
if [ "$compileMode" -eq 2 ]; then
	echo "pkill -9 -u $UID server" >> "$runnerScript"   #kill client and server if they exist       
	echo "pkill -9 -u $UID client" >> "$runnerScript"   #and are from user
	echo "cat <&0 > tmp" >> "$runnerScript"
	echo "$workdir/$serverExec < tmp 2>&1 &" >> "$runnerScript"
	echo "sleep 0.1" >> "$runnerScript"
	echo "$workdir/$clientExec < tmp 2>&1" >> "$runnerScript"
	echo "rm tmp" >> "$runnerScript"
#	echo "sleep 0.1" >> "$runnerScript"     
	echo "pkill -9 -u $UID server" >> "$runnerScript"               
	echo "exit 0" >> "$runnerScript"   #avoid return error if pkill does not find process
else
	echo "$workdir/$exec 2>&1" >> "$runnerScript"
fi

chmod 750 "$runnerScript"



java/myjava50.sh
#!/bin/bash
# 
# This script compiles java source code and run it.
#
#    a) if sole java file compile it and run it (file must have class with main function)
#
#    b) if zipped java files uncompresses it and compile all java files
#         run the class where main is specified (only one class can have a main file)
#
#    c) if specified in context Environment a client program, copies it to submission directory 
#      i) if a main function exists in this file it is ran
#     ii) also aditional classes to form a small API can be implemented in this file
#
#    Call:
#
#          myjava.sh   <mooshak dir> <submission_file> <extension> <problem_environment>
#
#    mooshak:
#          
#/bin/bash myjava.sh   $home         $file             $extension  $environment
#
#
#    cmd line
#
#	   myjava.sh   /home/mooshak/   src.zip           .zip     data/contests/POO1920/problems/P3/Environment
#
#
#	To run junit5 Mooshak "contest language" must be configure to:
#				MaxCompFork	10 (or 100)
#				MaxExecFork	10 (or 100)
#
#	And uncomment junit5 compile an run lines below
#
# Helder Daniel						
# hdaniel@ualg.pt
# September 2012
# April 2021

xtraZipDir=__MACOSX
extraCodeName="MooshakTester"
javaExt="java"
runnerScript="run.sh"
dfltWorkdir="src"

#Change for desired mode
username="mooshak"   # instdir: as in /home/$username/contrib (...)
junit="5"	      # junit 4 or 5
jdk="/usr"          #"/usr" or for local "/home/$username/jdk-15.0.2"
#jdk="/home/$username/jdk-15.0.2"
 
#get cur directory name
function curDir {
	result=$(basename $(pwd))
	echo $result
}


#get last created directory name
function newestDir {
	#remove extra Zip dir (MACOSX)
	if [ -d "$xtraZipDir" ]; then
		rm -r $xtraZipDir
	fi
	#get newest dir entry data
	result=$(ls -lF --sort=time | grep / | head -n1)

	#extract dir name
	echo ${result##*' '} | cut -d/ -f1

}


#Find file with main function
function findMainClass {
local workdir=$1

	#get 1st macth only
	result=$(grep -r "void[ ]*main" $workdir | head -n1)
	echo $result | cut -d: -f1 | cut -d. -f1
}


#get submitted code and unpack it if needed
function getSubmission {
local extension=$1
local file=$2

case "$extension" in
	.java)
	#Avoid error in creating dir in reevaluation
	if [ ! -d $dfltWorkdir ]; then
		/bin/mkdir $dfltWorkdir
	fi
	/bin/cp $file $dfltWorkdir
	;;

	.gz)
	#/bin/echo "tar.gz"
	#Send output of unzip to /dev/null to avoid returning it to mooshak.
	#Mooshak requires no output or consider an error
	/bin/tar xmzf $file > /dev/null
	;;

	.zip)
	#/bin/echo "zip"
	#Send output of unzip to /dev/null to avoid returning it to mooshak.
	#Mooshak requires no output or consider an error
	/usr/bin/unzip -o $file > /dev/null
	;;

	*)
	exit 1;
	;;
esac
}


#Add extra source code if it is defined in problem environment (from Mooshak GUI)
function addExtraCode {
local environment=$1
local workdir=$2
local extraCodeName=$3
local javaExt=$4

if [ -f "$environment" ]; then
	extraCode="$workdir/$extraCodeName.$javaExt"

	#Store extra source code in working dir with file name <submission dir>.java
	cp "$environment" "$extraCode"

	#If root package of submission not the default: <src>, set package 
	#in extracode file to workdir (root package name of submission)
	if [ "$workdir" != "src" ]; then
		echo "package $workdir;" | cat - $extraCode > temp && mv temp $extraCode
	fi

	#Search for junit4 or 5 in "$extraCode" 
	#junit4: "org.junit.Test"
	#junit5: "org.junit.jupiter.api.Test"
	#This ways detect junit4 or 5
	junit=$(grep "org.junit" "$extraCode" | wc -l)
	if [ $junit -gt 0 ]; then
		echo "junit"
	else
		#Search for a class $extraCodeName which should have main()
		result=$(grep -w "class[ ]*$extraCodeName" $extraCode)
		echo $result
	fi
fi
}



######################
#
# Main prg
#
######################

#Get script parameters
home="$1"
file="$2"
extension="$3"
environment="$home/$4"

#if revaluating DO NOT recreate files
#search for a report
#if [ -f "1.html" ]; then
#        exit 0
#fi

#get submission directory
curdir=$(curDir)
fullcurdir=$(pwd)

#if revaluating DO NOT recreate only SOURCE files
#but recreate run.sh.
#This allows to test changing the $ENVIRONMET Extra source code
if [ ! -f "1.html" ]; then
	#Unpack files if needed
	getSubmission $extension $file
fi

#Find unpacking/work directory
workdir=$(newestDir)

#Find file with main in submitted code
mainClass=$(findMainClass $workdir)
mainClassSoloName=$(echo ${mainClass##*'/'})
mainClassName=$(echo $mainClass | cut -d/ -f2-)

#Add extra source code if it is defined in problem environment (from Mooshak GUI)
#If it has a main() function set to run it instead of submitted main() function
result=$(addExtraCode $environment $workdir $extraCodeName $javaExt)
if [ "$result" != "" ]; then

	if [ "$result" != "junit" ]; then
		mainClassName=$extraCodeName
	fi
fi


#compile
find $workdir -name *.java > sources.txt

#This way allows that anyone submit junit 4/5 files too
#junit 4
if [ "$junit" = "4" ]; then
	$jdk/bin/javac -J-Dfile.encoding=UTF-8 -cp .:/home/$username/contrib/junit-4.12.jar @sources.txt -Xlint:-unchecked
fi
#junit 5
#(NEED fork>0 to run in safeexec)
if [ "$junit" = "5" ]; then
	$jdk/bin/javac -J-Dfile.encoding=UTF-8 -cp .:/home/$username/contrib/junit-platform-console-standalone-1.6.0.jar @sources.txt -Xlint #:-unchecked
fi

rm sources.txt


#Create runner script $runnerScript

#For alone files in zip
if [ "$workdir" = "" ]; then
	workdir="."
fi

echo "#!/bin/bash" > "$runnerScript"
if [ "$result" != "junit" ]; then
	#echo "$jdk/bin/java -cp $workdir $package$mainClassName 2>&1" >> "$runnerScript"
	echo "$jdk/bin/java -cp $workdir $mainClassName" >> "$runnerScript"
else
	#junit 4
	#Remove test time (which is variable) and last line
	#Not needed in below soultion, cause just prints error lines
	if [ "$junit" = "4" ]; then
		echo "$jdk/bin/java -cp .:/home/$username/contrib/junit-4.12.jar:/home/$username/contrib/hamcrest-core-1.3.jar:$workdir org.junit.runner.JUnitCore MooshakTester | grep 'Mooshak\|expected' | grep '^[^0-9]' | cut -d' ' -f2- 2>&1" >> "$runnerScript"
	fi

	#junit 5
	#(NEED fork>0 to run in safeexec)
	if [ "$junit" = "5" ]; then
		echo "$jdk/bin/java -jar /home/$username/contrib/junit-platform-console-standalone-1.6.0.jar --class-path $fullcurdir/$workdir --scan-class-path 2>&1" >> "$runnerScript"
	fi
fi

chmod 750 "$runnerScript"


